File: Solutions/Problem6/solution6.cpp
#include <iostream>
#include <string>
#include <cctype>

// User input
bool isPalindrome(const std::string& str) {
    int left = 0;
    int right = str.length() - 1;
    
    while (left < right) {
        while (left < right && !std::isalnum(str[left])) left++;
        while (left < right && !std::isalnum(str[right])) right--;
        
        if (std::tolower(str[left]) != std::tolower(str[right])) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}
// End of user input

class Checker {
public:
    static void check(bool (*func)(const std::string&), const std::string& input, bool expected) {
        bool result = func(input);
        if (result == expected) {
            std::cout << "Test Case Passed: Result = " << std::boolalpha << result 
                      << ", Expected = " << expected << std::endl;
        } else {
            std::cout << "Failed Test Case: Result = " << std::boolalpha << result 
                      << ", Expected = " << expected << std::endl;
        }
    }
};

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <string> <expected>" << std::endl;
        return 1;
    }

    std::string input = argv[1];
    bool expected = (std::string(argv[2]) == "1" || std::string(argv[2]) == "true");

    Checker::check(isPalindrome, input, expected);
    return 0;
} 

File: Solutions/Problem6/Solution6.java
public class Solution6 {
    // User input
    public static boolean isPalindrome(String str) {
        int left = 0;
        int right = str.length() - 1;
        
        while (left < right) {
            while (left < right && !Character.isLetterOrDigit(str.charAt(left))) left++;
            while (left < right && !Character.isLetterOrDigit(str.charAt(right))) right--;
            
            if (Character.toLowerCase(str.charAt(left)) != 
                Character.toLowerCase(str.charAt(right))) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
    // End of user input

    static class Checker {
        public static void check(String input, boolean expected) {
            boolean result = isPalindrome(input);
            if (result == expected) {
                System.out.printf("Test Case Passed: Result = %b, Expected = %b%n", 
                                result, expected);
            } else {
                System.out.printf("Failed Test Case: Result = %b, Expected = %b%n", 
                                result, expected);
            }
        }
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.printf("Usage: java Solution6 <string> <expected>%n");
            System.exit(1);
        }

        String input = args[0];
        boolean expected = Boolean.parseBoolean(args[1]);

        Checker.check(input, expected);
    }
} 